frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -0,0 +1,675 @@
+        #[pallet::weight(T::WeightInfo::bond())]
+        pub fn bond(
+            origin: OriginFor<T>,
+            controller: <T::Lookup as StaticLookup>::Source,
+            #[pallet::compact] value: BalanceOf<T>,
+            payee: RewardDestination<T::AccountId>,
+        ) -> DispatchResultWithPostInfo {
+            let stash = ensure_signed(origin)?;
+            ensure!(
+                !Bonded::<T>::contains_key(&stash),
+                Error::<T>::AlreadyBonded
+            );
+
+            let controller = T::Lookup::lookup(controller)?;
+            ensure!(
+                !Ledger::<T>::contains_key(&controller),
+                Error::<T>::AlreadyPaired
+            );
+
+            // reject a bond which is considered to be dust
+            ensure!(
+                value >= T::Currency::minimum_balance(),
+                Error::<T>::InsufficientValue
+            );
+
+            let stash_balance = T::Currency::free_balance(&stash);
+            ensure!(!stash_balance.is_zero(), Error::<T>::InsufficientValue);
+
+            Bonded::<T>::insert(&stash, &controller);
+            Payee::<T>::insert(&stash, payee);
+
+            Self::deposit_event(Event::Bonded(stash.clone(), stash_balance.clone()));
+
+            let value = value.min(stash_balance);
+            let ledger = StakingLedger {
+                stash,
+                total: value,
+                active: value,
+                unlocking: vec![],
+                last_reward: 0, // T::EraFinder::current() TODO
+            };
+            Self::update_ledger(&controller, &ledger);
+
+            Ok(().into())
+        }
+
+        /// Add some extra amount that have appeared in the stash `free_balance` into the balance up
+        /// for staking.
+        ///
+        /// Use this if there are additional funds in your stash account that you wish to bond.
+        /// Unlike [`bond`] or [`unbond`] this function does not impose any limitation on the amount
+        /// that can be added.
+        ///
+        /// The dispatch origin for this call must be _Signed_ by the stash, not the controller.
+        ///
+        /// # <weight>
+        /// - Independent of the arguments. Insignificant complexity.
+        /// - O(1).
+        /// - One DB entry.
+        /// # </weight>
+        #[pallet::weight(T::WeightInfo::bond_extra())]
+        pub fn bond_extra(
+            origin: OriginFor<T>,
+            #[pallet::compact] max_additional: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let stash = ensure_signed(origin)?;
+            ensure!(Bonded::<T>::contains_key(&stash), Error::<T>::NotStash);
+
+            let controller = Self::bonded(&stash).ok_or(Error::<T>::NotStash)?;
+            let mut ledger = Self::ledger(&controller).ok_or(Error::<T>::NotController)?;
+
+            let stash_balance = T::Currency::free_balance(&stash);
+
+            if let Some(extra) = stash_balance.checked_sub(&ledger.total) {
```

Comment:
@Maar-io this part confuses me a bit but I'm not familiar with financial terminology so this is expected.

How can currency be 'free' if it is locked?
Or in another words, if you bond certain amount of tokens, how can they be free?

I'm searching for more explanation online but would appreciate a short description if possible.

Reply:
if funds are bonded they are not free.
https://substrate.dev/rustdocs/latest/frame_support/traits/tokens/currency/trait.Currency.html#tymethod.free_balance

Reply:
I cannot remember we resolved this (during our sync) and I still have this under TODO.


What you wrote makes sense and I also understood it like that.

But that logic isn't reflected in the code above.
If that were true, you wouldn't need to subtract ledger.total from free_balance of the stash.

So what is up with that?


In general, I think it would be good to add some description of how something was resolved so we don't have situations like this.